app/services/SessionService.scala

```diff
@@ -17,26 +17,88 @@
     }
   }
 
-  def set(key: String, value: String, result: Result)(implicit ec: ExecutionContext, request: RequestHeader): Future[Either[Throwable, Result]] = {
+  def getMongo(journeyType: String)(implicit hc: HeaderCarrier, ec: ExecutionContext): Future[Either[Throwable, Option[UIJourneySessionData]]] = {
+    uiJourneySessionDataRepository.get(hc.sessionId.get.value, journeyType) map {
+      case Some(data: UIJourneySessionData) =>
+        Right(Some(data))
+      case None => Right(None)
+    }
+  }
+
+  def createSession(journeyType: String)(implicit hc: HeaderCarrier, ec: ExecutionContext): Future[Boolean] = {
+    setMongoData(UIJourneySessionData(hc.sessionId.get.value, journeyType, None))
+  }
+
+  private def getKeyFromObject(objectOpt: Option[Any], key: String): Option[String] = {
+    objectOpt.fold(Option(""))(
+      obj => {
+        val field = obj.getClass.getDeclaredField(key)
+        field.setAccessible(true)
+        val value: Option[String] = field.get(obj).asInstanceOf[Option[String]]
+        value
+      }
+    )
+  }
+
+  def getMongoKey(key: String, journeyType: JourneyType)
+                 (implicit hc: HeaderCarrier, ec: ExecutionContext): Future[Either[Throwable, Option[String]]] = {
+    uiJourneySessionDataRepository.get(hc.sessionId.get.value, journeyType.toString) map {
+      case Some(data: UIJourneySessionData) =>
+        journeyType.operation match {
+          case Add => Right(getKeyFromObject(data.addIncomeSourceData, key))
+          case Manage => Right(getKeyFromObject(data.manageIncomeSourceData, key))
+          case Cease => Right(getKeyFromObject(data.ceaseIncomeSourceData, key))
+        }
+      case None => Right(None)
+    }
+  }
+
+  def set(key: String, value: String, result: Result)
+         (implicit ec: ExecutionContext, request: RequestHeader): Future[Either[Throwable, Result]] = {
     Future {
       Right(result.addingToSession(key -> value))
     }
   }
 
+  def setMongoData(uiJourneySessionData: UIJourneySessionData)
+                  (implicit hc: HeaderCarrier, ec: ExecutionContext): Future[Boolean] = {
+    uiJourneySessionDataRepository.set(uiJourneySessionData)
+  }
+
+  def setMongoKey(key: String, value: String, journeyType: JourneyType)
+                 (implicit hc: HeaderCarrier, ec: ExecutionContext): Future[Either[Throwable, Boolean]] = {
+    val uiJourneySessionData = UIJourneySessionData(hc.sessionId.get.value, journeyType.toString, None)
+    val jsonAccessorPath = journeyType.operation match {
+      case Add => AddIncomeSourceData.getJSONKeyPath(key)
+      case Manage => ManageIncomeSourceData.getJSONKeyPath(key)
+      case Cease => CeaseIncomeSourceData.getJSONKeyPath(key)
+    }
+    uiJourneySessionDataRepository.updateData(uiJourneySessionData, jsonAccessorPath, value).map(
+      result => result.wasAcknowledged() match {
+        case true => Right(true)
+        case false => Left(new Error("Mongo Save data operation was not acknowledged"))
```

Comment:
I think this should be an exception rather than an error as it's non-fatal, maybe a RuntimeException?
https://stackoverflow.com/questions/5813614/what-is-difference-between-errors-and-exceptions#:~:text=An%20Error%20%22indicates%20serious%20problems,should%20not%20try%20to%20catch.%22&text=An%20Exception%20%22indicates%20conditions%20that,Exception%20classes%20are%20checked%20exceptions.

Reply:
I think Error and Exception here are interchangeable, its basically the Throwable aspect that we really need. That stackoverflow thread is highly opinionated..

Reply:
<img width="2479" alt="Screenshot 2023-09-28 at 17 04 40" src="https://github.com/hmrc/income-tax-view-change-frontend/assets/3528155/10b165ed-655a-491f-b9f5-201c7aa1073e">
Error - is a branch of JVM level exception, where is 
Exception is program level exceptions